libbeat/publisher/pipeline/pipeline.go

```diff
@@ -386,3 +389,26 @@ func (p *Pipeline) createEventProcessing(cfg beat.ProcessingConfig, noPublish bo
 func (p *Pipeline) OutputReloader() OutputReloader {
 	return p.outputController
 }
+
+// Parses the given config and returns a memqueue.Settings or
+// diskqueue.Settings based on it. This helper exists to frontload config
+// parsing errors: if there is an error in the queue config, we want it to
+// show up as fatal during initialization, even if the queue itself isn't
+// created until later.
+func queueSettingsForUserConfig(queueType string, userConfig *conf.C, inQueueSize int) (interface{}, error) {
```

Comment:
That bare interface really bugs me, even if I understand why we ended up with it. At very least I kinda wonder if we could have some kind of named interface or something, just for the sake of making the API itself look less opaque.

Reply:
It bugs me too and it wasn't my first choice. The constraint that led me here is that we need to be able to fail during initialization if we get an invalid queue config up front, therefore we need to call `SettingsFromConfig` for the appropriate queue since they all have incompatible settings, but we also need that to be overridable by something the output can create.
- Alternative: We could panic later when we realize the config is invalid. I don't like this one.
- Alternative: We could report an error when we see it but fall back on the default memory queue. (I'm open to this one but in the end I decided quitting on initialization was preferable.)

The settings alone aren't enough to create the queue, we also need the logger (which we could get from other sources in a pinch) and the queue's global acknowledgment callback (which we can't) so we can't have the output actually _create_ the queue. On the other hand, just accepting a queue type / `*conf.C` is also awkward: the proxy queue isn't user-configurable, so a natural future choice like passing along a `queue` subtree from an `output` configuration would expose that internal type. (Besides which, to me an `interface{}` that is a placeholder for 3 concrete types is still a lot easier to work with than a `*conf.C` that could contain literally anything).

The old way of handling this tension was to use factories, which gives thing a nice name but makes the internals too opaque: a factory that just took the callback and created a queue would work, but there would be no way to know what type of queue we'd created, which we currently report.

Which... having talked myself into a corner, maybe we just need to add a `QueueType()` to the queue interface, and have both the pipeline initialization and the outputs give a queue factory for this field? At this point it could be a `func(*logp.Logger, func(eventCount int)) (queue.Queue, error)` which is more consistent than the old types, and we could still fall back on the memory queue on error (knowing that an honest queue creation can only fail for the disk queue). I dislike losing access to the settings objects but right now we don't actually need them so this might make sense, thoughts?

Reply:
Ok, I switched to a version using factories, see if you feel better about this one ^^